test/models/meshgraphnet/test_meshgraphnet_snmg.py

```diff
@@ -186,22 +198,20 @@ def run_test_distributed_meshgraphnet(rank, world_size, dtype, partition_scheme)
     loss.backward()
 
     out_multi_gpu = model_multi_gpu(nfeat_multi_gpu, efeat_multi_gpu, graph_multi_gpu)
```

Comment:
Can you help me understand how this multi gpu model is working? Looks like I'm missing something about how the partitioned graph gets set in the meshgraphnet model. I don't think I see any attributes of `model_multi_gpu` being set after construction and the constructor doesn't take in any graph data right?

Reply:
You are right. MeshGraphNet in its current design takes in the graph in the forward method. `model_multi_gpu` thus is more or less just the same model as in the single-gpu case on each rank with the only difference that you call `mark_module_as_shared` to register the necessary backward hooks.

The other difference to the single-gpu case is that you specify this "graph" object explicitly with a `graph_partition` and the process group.

```    
graph_multi_gpu = CuGraphCSC(
        offsets.to(manager.device),
        indices.to(manager.device),
        num_nodes,
        num_nodes,
        partition_size=world_size,
        partition_group_name="graph_partition",
        graph_partition=graph_partition,
    )
```

This example actually existed before this PR. I just improved a few minor things here and there after discovering initial bugs in the partitioning logic.

The distributed GraphCastNet on its own actually also doesn't have too many differences from its single-gpu model. As this model "own" more information (e.g. about creating the graph structures in the constructor, converting the layout of input/outputs within the forward pass, etc..), you have to move more of these setups into the model.